### DISCUSSION TITLE
jq (specifically jq) is incredibly slow

### DISCUSSION THREAD

**Discussion by Readf0x** [elkowar/eww#1018] [Q&A]
I have tested this multiple ways, using jq in a command causes a delay so large the widget becomes near unusable. I have my workspace widget set up like so:
```
(deflisten workspaces "hyprland-workspaces _ 2>/dev/null | jq -cMra 'sort_by(.name | try(tonumber))'")

(defwidget workspaces []
  (box
    :class "workspaces"
    :orientation "h"
    :space-evenly false
    (box
      :class "variable-buttons"
      :space-evenly false
      (for workspace in workspaces
        (button
          :onclick "hyprctl dispatch workspace ${workspace.name}"
          :class "${workspace.id} ${workspace.active ? 'active' : ''}"
          "${workspace.name =~ [0-9] ? workspace.name : ''}${workspace.name == 'info' ? '' : ''}${workspace.name == 'music' ? '󰝚' : ''}${workspace.name == 'video' ? '' : ''}${workspace.name == 'discord' ? '󰙯' : ''}"
        )
      )
    )
    (button
      :class "dropdown"
      :onclick "hyprctl dispatch togglespecialworkspace dropdown"
      ""
    )
  )
)
```

for some reason unbeknownst to me, it's like the command is only outputting once every 30 seconds or so. Just to make sure it's not me, I tried not actually filtering anything in jq and just running the json through it, like this:
```
(deflisten workspaces "~/.cargo/bin/hyprland-workspaces _ 2>/dev/null | jq")
```
The same problem occurs.
I tested it in the terminal and it works fine. The widget functions perfectly fine if I simply don't pipe it into jq. Any pointers?

**Comment by YellowOnion** [comment]
There's different possibilities, 

1. hyprland-workspaces is doing something weird with output, causing buffering to not be flushed at correct time.
2. jq is failing in the top and bottom for difference reasons, try `jq -c` for minimal example, raw output (`-r`) makes no sense if you're trying to output json either.
3. don't pipe stderr to null, if you're having errors they'll show up in `eww logs`.
4. You have your listen variable and widget with the same name, this could be causing issues.

Also replace the chained ${}${}...${} code with chained ternaries, like this, it should be a lot easier to read:
```
"${condition1 ? value1
    : condition2 ? value2
    : condition3 ? value3
    : value4}"
```

It would also be a lot easier to debug if you provided output of `eww logs`, example json from `hyprland-workspaces` (I can't find this tool online?).

**Reply by Readf0x** [reply]
`hyprland-workspaces` was a tool often used with eww, it was popular on arch. Idk if it's still around anymore but unfortunately I no longer use eww, or even have my original dots. The issue probably doesn't occur anymore, especially seeing as I was the only one with this problem an entire year ago. Probably user error or something.

**Comment by hypernova7** [comment]
Have you tried [gojq](https://github.com/itchyny/gojq)?
